### PR TITLE
Custom taints and toleration node operation

### PR DESCRIPTION
**Pull request by vkathole**

### CODE REVIEW

File: ocs_ci/ocs/resources/pod.py [hunk 1]
```diff
@@ -3719,3 +3719,30 @@ def verify_md5sum_on_pod_files(pods_for_integrity_check, pod_file_name):
             f"matches with the original md5sum"
         )
     logger.info("Data integrity check passed on all pods")
+
+
+def check_toleration_on_subscriptions(toleration_key=constants.TOLERATION_KEY):
+    """
+    Function to check toleration on subscriptions
+    Args:
+        toleration_key (str): The toleration key to check
```
Comment (Shrivaibavi) [inline ocs_ci/ocs/resources/pod.py:3728]: Need a space after this line

File: tests/functional/z_cluster/nodes/test_non_ocs_taint_and_toleration.py [hunk 2]
```diff
@@ -227,3 +369,140 @@ def test_non_ocs_taint_and_tolerations(self):
...
+    @skipif_ocs_version("<=4.15")
+    def test_custom_taint_node_operations(self, nodes):
+        """
+        Test runs the following steps
+        1. Add toleration in storagecluster CR and odf-operator subscription.
+        2. Verify toleration applied in ODF subscription is reflecting
+           for other subscriptions, Ceph and nooba components or not(Check all 7 subscriptions).
+        3. Apply custom taint to all nodes.
+        4. Verify the pods in all nodes are running as per taints applied.
+        5. Reboot one of the nodes and check toleration on all odf pods on that node.
```
Comment (Shrivaibavi) [inline tests/functional/z_cluster/nodes/test_non_ocs_taint_and_toleration.py:382]: missing step 6

File: tests/functional/z_cluster/nodes/test_non_ocs_taint_and_toleration.py [hunk 3]
```diff
@@ -227,3 +369,140 @@ def test_non_ocs_taint_and_tolerations(self):
...
+            cephcluster = CephClusterExternal()
+            cephcluster.cluster_health_check()
+        else:
+            self.sanity_helpers.health_check()
+
+        # Get the node list
+        node = get_nodes("worker", num_of_nodes=1)
+
+        # Reboot one master nodes
+        nodes.restart_nodes(node, wait=False)
```
Comment (Shrivaibavi) [inline tests/functional/z_cluster/nodes/test_non_ocs_taint_and_toleration.py:479]: shouldn't the wait had to be true ?

File: tests/functional/z_cluster/nodes/test_non_ocs_taint_and_toleration.py [hunk 4]
```diff
@@ -227,3 +369,140 @@ def test_non_ocs_taint_and_tolerations(self):
...
+        if config.DEPLOYMENT["external_mode"]:
+            cephcluster = CephClusterExternal()
+            cephcluster.cluster_health_check()
+        else:
+            self.sanity_helpers.health_check()
+
+        # Get the node list
+        node = get_nodes("worker", num_of_nodes=1)
+
+        # Reboot one master nodes
```
Comment (Shrivaibavi) [inline tests/functional/z_cluster/nodes/test_non_ocs_taint_and_toleration.py:478]: Step 5 says worker node and the comment # contradicts

File: tests/functional/z_cluster/nodes/test_non_ocs_taint_and_toleration.py [hunk 5]
```diff
@@ -227,3 +369,140 @@ def test_non_ocs_taint_and_tolerations(self):
...
+        nodes.restart_nodes(node, wait=False)
+
+        # Wait some time after rebooting master
```
Comment (Shrivaibavi) [inline tests/functional/z_cluster/nodes/test_non_ocs_taint_and_toleration.py:481]: is it master or worker ?

File: tests/functional/z_cluster/nodes/test_non_ocs_taint_and_toleration.py [hunk 6]
```diff
@@ -227,3 +369,140 @@ def test_non_ocs_taint_and_tolerations(self):
...
+        waiting_time = 40
+        logger.info(f"Waiting {waiting_time} seconds...")
+        time.sleep(waiting_time)
+
+        # Validate all nodes and services are in READY state and up
```
Comment (Shrivaibavi) [inline tests/functional/z_cluster/nodes/test_non_ocs_taint_and_toleration.py:486]: Test step says : 5. Reboot one of the nodes and check toleration on all odf pods on that node.

where are we checking toleration on odf pods after reboots ?

File: tests/functional/z_cluster/nodes/test_non_ocs_taint_and_toleration.py [hunk 7]
```diff
@@ -35,6 +44,139 @@
...
+    """
+    Delete an osd node, and create a new one to replace it
+
+    Args:
+        osd_node_name (str): The osd node name to delete
+
+    """
+    new_node_name = None
+    old_osd_ids = node.get_node_osd_ids(osd_node_name)
+
```
Comment (Shrivaibavi) [inline tests/functional/z_cluster/nodes/test_non_ocs_taint_and_toleration.py:139]: remove the extra line

File: tests/functional/z_cluster/nodes/test_non_ocs_taint_and_toleration.py [hunk 8]
```diff
@@ -227,3 +369,140 @@ def test_non_ocs_taint_and_tolerations(self):
...
+        2. Verify toleration applied in ODF subscription is reflecting
+           for other subscriptions, Ceph and nooba components or not(Check all 7 subscriptions).
+        3. Apply custom taint to all nodes.
+        4. Verify the pods in all nodes are running as per taints applied.
+        5. Reboot one of the nodes and check toleration on all odf pods on that node.
+        7.Replace one of the nodes and check all odf pods on that node are running.
+        """
+
+        logger.info("Taint all nodes with custom taint")
+        ocs_nodes = get_worker_nodes()
```
Comment (PrasadDesala) [inline tests/functional/z_cluster/nodes/test_non_ocs_taint_and_toleration.py:387]: this fun returns all the worker nodes, should we taint only on odf pod running nodes?

File: tests/functional/z_cluster/nodes/test_non_ocs_taint_and_toleration.py [hunk 9]
```diff
@@ -227,3 +369,140 @@ def test_non_ocs_taint_and_tolerations(self):
...
+        taint_nodes(nodes=ocs_nodes, taint_label="xyz=true:NoSchedule")
+
+        resource_name = constants.DEFAULT_CLUSTERNAME
+        if config.DEPLOYMENT["external_mode"]:
+            resource_name = constants.DEFAULT_CLUSTERNAME_EXTERNAL_MODE
+
+        logger.info("Add tolerations to storagecluster")
```
Comment (PrasadDesala) [inline tests/functional/z_cluster/nodes/test_non_ocs_taint_and_toleration.py:394]: the steps looks almost similar to the existing non ocs taints test, what has changed?

File: tests/functional/z_cluster/nodes/test_non_ocs_taint_and_toleration.py [hunk 10]
```diff
@@ -35,6 +44,139 @@
...
+def delete_and_create_osd_node(osd_node_name):
+    """
+    Delete an osd node, and create a new one to replace it
+
+    Args:
+        osd_node_name (str): The osd node name to delete
+
+    """
+    new_node_name = None
+    old_osd_ids = node.get_node_osd_ids(osd_node_name)
```
Comment (PrasadDesala) [inline tests/functional/z_cluster/nodes/test_non_ocs_taint_and_toleration.py:138]: you are duplicating the node and device replacement func, we cannot reuse the existing func?

File: tests/functional/z_cluster/nodes/test_non_ocs_taint_and_toleration.py [hunk 11]
```diff
@@ -75,15 +86,48 @@ def finalizer():
...
+            logger.info("Remove tolerations from storagecluster")
+            storagecluster_obj = ocp.OCP(
+                resource_name=resource_name,
+                namespace=config.ENV_DATA["cluster_namespace"],
+                kind=constants.STORAGECLUSTER,
+            )
+            params = '[{"op": "remove", "path": "/spec/placement"},]'
+            storagecluster_obj.patch(params=params, format_type="json")
+
+            logger.info("Remove tolerations to the subscription")
```
Comment (Shrivaibavi) [inline tests/functional/z_cluster/nodes/test_non_ocs_taint_and_toleration.py:102]:
```suggestion
            logger.info("Remove tolerations from the subscriptions")
```

File: tests/functional/z_cluster/nodes/test_non_ocs_taint_and_toleration.py [hunk 12]
```diff
@@ -75,15 +86,48 @@ def finalizer():
...
+            )
+            for sub in sub_list:
+                subscription_data = sub_obj.get(resource_name=sub)
+                if "config" in subscription_data.get("spec", {}):
+                    params = '[{"op": "remove", "path": "/spec/config"}]'
+                    sub_obj.patch(resource_name=sub, params=params, format_type="json")
+            time.sleep(180)
+            assert wait_for_pods_to_be_running(
+                timeout=900, sleep=15
+            ), "some of the pods didn't came up running"
```
Comment (Shrivaibavi) [inline tests/functional/z_cluster/nodes/test_non_ocs_taint_and_toleration.py:117]:
```suggestion
            ), "Few pods failed to reach the desired running state"
```

File: tests/functional/z_cluster/nodes/test_non_ocs_taint_and_toleration.py [hunk 13]
```diff
@@ -75,15 +86,48 @@ def finalizer():
...
+            sub_obj = ocp.OCP(
+                namespace=config.ENV_DATA["cluster_namespace"],
+                kind=constants.SUBSCRIPTION,
+            )
+            for sub in sub_list:
+                subscription_data = sub_obj.get(resource_name=sub)
+                if "config" in subscription_data.get("spec", {}):
+                    params = '[{"op": "remove", "path": "/spec/config"}]'
+                    sub_obj.patch(resource_name=sub, params=params, format_type="json")
+            time.sleep(180)
```
Comment (Shrivaibavi) [inline tests/functional/z_cluster/nodes/test_non_ocs_taint_and_toleration.py:114]: are we not supposed to remove the tolerations from the rook-ceph operator configmap and ocsinitializations.ocs.openshift.io too ??

File: tests/functional/z_cluster/nodes/test_non_ocs_taint_and_toleration.py [hunk 14]
```diff
@@ -255,3 +299,145 @@ def test_non_ocs_taint_and_tolerations(self):
                 resource_count=count * replica_count,
             ), "New OSDs failed to reach running state"
             check_ceph_health_after_add_capacity(ceph_rebalance_timeout=2500)
+
+        # Reboot one of the nodes
```
[the remark this answers is not in the document]
Comment (Shrivaibavi) [reply tests/functional/z_cluster/nodes/test_non_ocs_taint_and_toleration.py:303]: +1

File: tests/functional/z_cluster/nodes/test_non_ocs_taint_and_toleration.py [hunk 15]
```diff
@@ -255,3 +299,145 @@ def test_non_ocs_taint_and_tolerations(self):
...
+        logger.info("Verifying All resources are Running and matches expected result")
+        retry((CommandFailed, TolerationNotFoundException), tries=5, delay=10,)(
+            check_toleration_on_pods
+        )(toleration_key="xyz")
+        self.sanity_helpers.health_check(tries=120)
+
+    def test_negative_custom_taint(self, nodes):
+        """
+        Test runs the following steps
+        1. Taint odf nodes with non-ocs taint
```
Comment (Shrivaibavi) [inline tests/functional/z_cluster/nodes/test_non_ocs_taint_and_toleration.py:349]:
```suggestion
        1. Taint odf worker nodes with non-ocs taint
```

File: tests/functional/z_cluster/nodes/test_non_ocs_taint_and_toleration.py [hunk 16]
```diff
@@ -255,3 +299,145 @@ def test_non_ocs_taint_and_tolerations(self):
...
+        pod_list = get_all_pods(
+            namespace=config.ENV_DATA["cluster_namespace"],
+            exclude_selector=True,
+        )
+        for pod in pod_list:
+            pod.delete(wait=False)
+
+        assert not wait_for_pods_to_be_running(
+            timeout=120, sleep=15
+        ), "Pods are running when they should not be."
```
Comment (Shrivaibavi) [inline tests/functional/z_cluster/nodes/test_non_ocs_taint_and_toleration.py:434]: are we expecting all pods to go in a bad state ?

Comment (Shrivaibavi) [reply tests/functional/z_cluster/nodes/test_non_ocs_taint_and_toleration.py:434]: I see we apply tolerations on storagecluster and subscription other than ODF, are we sure all pods will not be running if the toleration is just not applied properly on sub when we are setting it properly on storagecluster ? Please check the scenario again. if we are setting the toleration properly on storagecluster few pods should be up and running.

File: tests/functional/z_cluster/nodes/test_non_ocs_taint_and_toleration.py [hunk 17]
```diff
@@ -255,3 +299,145 @@ def test_non_ocs_taint_and_tolerations(self):
...
+
+        logger.info(
+            "Check custom toleration on all newly created pods under openshift-storage"
```
Comment (Shrivaibavi) [inline tests/functional/z_cluster/nodes/test_non_ocs_taint_and_toleration.py:437]:
```suggestion
            "Validate custom toleration not found on all newly created pods in openshift-storage"
```

**Review by am-agrawa — APPROVED** [review]

File: tests/functional/z_cluster/nodes/test_non_ocs_taint_and_toleration.py [hunk 18]
```diff
@@ -75,25 +84,47 @@ def finalizer():
...
+                    sub_obj.patch(resource_name=sub, params=params, format_type="json")
+            time.sleep(180)
+            assert wait_for_pods_to_be_running(
+                timeout=900, sleep=15
+            ), "Few pods failed to reach the desired running state"
+
         request.addfinalizer(finalizer)
 
-    def test_non_ocs_taint_and_tolerations(self):
+    def apply_custom_taint_and_toleration(self):
```
Comment (PrasadDesala) [inline tests/functional/z_cluster/nodes/test_non_ocs_taint_and_toleration.py:119]: should we move it to some common place outside test? 
You can have an arg with taint_label to this function

File: tests/functional/z_cluster/nodes/test_non_ocs_taint_and_toleration.py [hunk 19]
```diff
@@ -197,28 +223,55 @@ def test_non_ocs_taint_and_tolerations(self):
...
+    def test_non_ocs_taint_and_tolerations(self, nodes):
+        """
+        Test runs the following steps
+        1. Taint odf nodes with non-ocs taint
+        2. Set tolerations on storagecluster, subscription, configmap and ocsinit
+        3. check tolerations on all subscription yaml.
+        4. Check toleration on all odf pods.
+        5. Add Capacity.
+        6. Reboot one of the nodes and check toleration on all odf pods on that node.
+        7. Replace one of the nodes and check all odf pods on that node are running.
```
Comment (PrasadDesala) [inline tests/functional/z_cluster/nodes/test_non_ocs_taint_and_toleration.py:248]: you are not doing these two steps in the test. please delete these steps

File: tests/functional/z_cluster/nodes/test_non_ocs_taint_and_toleration.py [hunk 20]
```diff
@@ -75,25 +84,47 @@ def finalizer():
...
 
-    def test_non_ocs_taint_and_tolerations(self):
+    def apply_custom_taint_and_toleration(self):
         """
-        Test runs the following steps
-        1. Taint ocs nodes with non-ocs taint
-        2. Set tolerations on storagecluster, subscription, configmap and ocsinit
-        3. Check toleration on all ocs pods.
-        4. Add Capacity
+        Apply custom taints and tolerations.
```
Comment (PrasadDesala) [inline tests/functional/z_cluster/nodes/test_non_ocs_taint_and_toleration.py:121]: Please add the detailed steps

File: tests/functional/z_cluster/nodes/test_non_ocs_taint_and_toleration.py [hunk 21]
```diff
@@ -197,28 +223,55 @@ def test_non_ocs_taint_and_tolerations(self):
...
+            timeout=900, sleep=15
+        ), "Few pods failed to reach the desired running state"
 
         logger.info(
             "Check non-ocs toleration on all newly created pods under openshift-storage NS"
         )
-        retry(CommandFailed, tries=5, delay=10,)(
+        retry((CommandFailed, TolerationNotFoundException), tries=10, delay=10,)(
             check_toleration_on_pods
         )(toleration_key="xyz")
```
Comment (PrasadDesala) [inline tests/functional/z_cluster/nodes/test_non_ocs_taint_and_toleration.py:276]: why are we checking this twice, above one in line 259

File: tests/functional/z_cluster/nodes/test_non_ocs_taint_and_toleration.py [hunk 22]
```diff
@@ -255,3 +308,198 @@ def test_non_ocs_taint_and_tolerations(self):
...
+        ), "Few pods failed to reach the desired running state"
+
+        logger.info(
+            "Check non-ocs toleration on all newly created pods under openshift-storage NS"
+        )
+        retry((CommandFailed, TolerationNotFoundException), tries=10, delay=10,)(
+            check_toleration_on_pods
+        )(toleration_key="xyz")
+        # Reboot one of the nodes
+        node = get_ocs_nodes(num_of_nodes=1)
```
Comment (PrasadDesala) [inline tests/functional/z_cluster/nodes/test_non_ocs_taint_and_toleration.py:344]: please get a random node

File: tests/functional/z_cluster/nodes/test_non_ocs_taint_and_toleration.py [hunk 23]
```diff
@@ -197,28 +223,55 @@ def test_non_ocs_taint_and_tolerations(self):
...
+        number_of_pods_before = len(
+            get_all_pods(namespace=config.ENV_DATA["cluster_namespace"])
+        )
+
+        logger.info("Apply custom taints and tolerations.")
+        self.apply_custom_taint_and_toleration()
+
+        retry((CommandFailed, TolerationNotFoundException), tries=5, delay=10,)(
+            check_toleration_on_subscriptions
+        )(toleration_key="xyz")
```
Comment (PrasadDesala) [inline tests/functional/z_cluster/nodes/test_non_ocs_taint_and_toleration.py:261]: any reason for specifically checking tolerations on subscriptions? check_toleration_on_pods is not enough?

File: ocs_ci/helpers/helpers.py [hunk 24]
```diff
@@ -5392,3 +5401,130 @@ def verify_performance_profile_change(perf_profile):
     ), f"Performance profile is not updated successfully to {perf_profile}"
     logger.info(f"Performance profile successfully got updated to {perf_profile} mode")
     return True
+
+
+def apply_custom_taint_and_toleration(taint_label):
```
Comment (PrasadDesala) [inline ocs_ci/helpers/helpers.py:5406]: you can have the custom taint label defined here in the arg

File: tests/functional/z_cluster/nodes/test_non_ocs_taint_and_toleration.py (hunk 16 above)
Comment (Shrivaibavi) [reply tests/functional/z_cluster/nodes/test_non_ocs_taint_and_toleration.py:408]: Change the assert message please

**Review by PrasadDesala — APPROVED** [review]

**Review by am-agrawa — APPROVED** [review]

**Pull request merged**